### PR TITLE
Misc

### LINKED ISSUE #106: COW violation in sodium_increment
In https://github.com/jedisct1/libsodium-php/blob/master/libsodium.c#L452 a string is modified directly, which might have rc>1 (i.e. a SEPARATE_STRING call is missing). Additionally the parameter should be accepted by reference.

### PR DESCRIPTION
**Pull request by nikic**
In the interest of getting the ball on bundling rolling, this PR
 * imports the migration to exceptions from #109
 * switches to SodiumException
 * drops argument information from these exceptions
 * drops the PHP 5 compatibility code
 * fixes #106 and a related pair of segfaults

### CODE REVIEW

**Comment by paragonie-scott** [comment]
The currently-failing test is a Travis timeout. LGTM.

File: libsodium.c [hunk 1]
```diff
@@ -516,18 +565,19 @@ PHP_FUNCTION(sodium_memcmp)
...
 
-    if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "ss",
+    if (zend_parse_parameters(ZEND_NUM_ARGS(), "ss",
                               &buf1, &len1,
                               &buf2, &len2) == FAILURE) {
         return;
     }
     if (len1 != len2) {
         RETURN_LONG(-1);
     } else if (len1 > SIZE_MAX) {
```
Comment (nikic) [inline libsodium.c]: This check no longer makes sense.

File: libsodium.c [hunk 2]
```diff
@@ -603,35 +658,40 @@ PHP_FUNCTION(sodium_crypto_secretbox)
...
+        zend_throw_exception(sodium_exception_ce, "arithmetic overflow", 0);
+        return;
     }
     ciphertext = zend_string_alloc((size_t) msg_len + crypto_secretbox_MACBYTES, 0);
     if (crypto_secretbox_easy((unsigned char *) ZSTR_VAL(ciphertext),
                               msg, (unsigned long long) msg_len,
                               nonce, key) != 0) {
         zend_string_free(ciphertext);
-        zend_error(E_RECOVERABLE_ERROR, "crypto_secretbox()");
+        zend_throw_exception(zend_ce_error, "crypto_secretbox(): internal error", 0);
```
Comment (nikic) [inline libsodium.c]: return missing here

File: libsodium.c [hunk 3]
```diff
@@ -723,25 +790,27 @@ PHP_FUNCTION(sodium_crypto_generichash_init)
...
         hash_len > crypto_generichash_BYTES_MAX) {
-        zend_error(E_RECOVERABLE_ERROR, "crypto_generichash_init(): unsupported output length");
+        zend_throw_exception(sodium_exception_ce, "crypto_generichash_init(): unsupported output length", 0);
+        return;
     }
     if (key_len != 0 &&
         (key_len < crypto_generichash_KEYBYTES_MIN ||
          key_len > crypto_generichash_KEYBYTES_MAX)) {
-        zend_error(E_RECOVERABLE_ERROR, "crypto_generichash_init(): unsupported key length");
+        zend_throw_exception(sodium_exception_ce, "crypto_generichash_init(): unsupported key length", 0);
```
Comment (nikic) [inline libsodium.c]: Missing return here.

File: libsodium.c [hunk 4]
```diff
@@ -1182,7 +1291,7 @@ PHP_FUNCTION(sodium_crypto_sign_seed_keypair)
                                  (unsigned char *) ZSTR_VAL(keypair),
                                  seed) != 0) {
         zend_string_free(keypair);
-        zend_error(E_RECOVERABLE_ERROR, "crypto_sign_seed_keypair()");
+        zend_throw_exception(zend_ce_error, "crypto_sign_seed_keypair(): internal error", 0);
```
Comment (nikic) [inline libsodium.c]: Missing return here.

File: libsodium.c [hunk 5]
```diff
@@ -1489,29 +1630,29 @@ PHP_FUNCTION(sodium_crypto_stream_xor)
...
-    if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "sss",
+    if (zend_parse_parameters(ZEND_NUM_ARGS(), "sss",
                               &msg, &msg_len,
                               &nonce, &nonce_len,
                               &key, &key_len) == FAILURE) {
         return;
     }
     if (nonce_len != crypto_stream_NONCEBYTES) {
-        zend_error(E_RECOVERABLE_ERROR, "nonce should be CRYPTO_STREAM_NONCEBYTES bytes");
+        zend_throw_exception(sodium_exception_ce, "nonce should be CRYPTO_STREAM_NONCEBYTES bytes", 0);
```
Comment (nikic) [inline libsodium.c]: Missing return here and on three following occurances.

File: libsodium.c [hunk 6]
```diff
@@ -1912,14 +2088,15 @@ PHP_FUNCTION(sodium_crypto_aead_chacha20poly1305_encrypt)
...
-        zend_error(E_RECOVERABLE_ERROR, "crypto_aead_chacha20poly1305_encrypt()");
+        zend_throw_exception(zend_ce_error, "crypto_aead_chacha20poly1305_encrypt(): internal error", 0);
+        return;
     }
-    if (ciphertext_real_len <= 0U || ciphertext_real_len >= STRSIZE_MAX ||
+    if (ciphertext_real_len <= 0U || ciphertext_real_len >= SIZE_MAX ||
         ciphertext_real_len > ciphertext_len) {
         zend_string_free(ciphertext);
-        zend_error(E_RECOVERABLE_ERROR, "arithmetic overflow");
+        zend_throw_exception(sodium_exception_ce, "arithmetic overflow", 0);
```
Comment (nikic) [inline libsodium.c]: Missing return.

File: libsodium.c [hunk 7]
```diff
@@ -1971,11 +2153,11 @@ PHP_FUNCTION(sodium_crypto_aead_chacha20poly1305_decrypt)
         zend_string_free(msg);
         RETURN_FALSE;
     }
-    if (msg_real_len >= STRSIZE_MAX || msg_real_len > msg_len) {
+    if (msg_real_len >= SIZE_MAX || msg_real_len > msg_len) {
         zend_string_free(msg);
-        zend_error(E_RECOVERABLE_ERROR, "arithmetic overflow");
+        zend_throw_exception(sodium_exception_ce, "arithmetic overflow", 0);
```
Comment (nikic) [inline libsodium.c]: Missing return

File: libsodium.c [hunk 8]
```diff
@@ -2315,22 +2527,24 @@ PHP_FUNCTION(sodium_crypto_sign_ed25519_pk_to_curve25519)
...
+                   "Ed25519 key should be CRYPTO_SIGN_PUBLICKEYBYTES long",
+                   0);
+        return;
     }
     ecdhkey = zend_string_alloc(crypto_sign_PUBLICKEYBYTES, 0);
 
     if (crypto_sign_ed25519_pk_to_curve25519((unsigned char *) ZSTR_VAL(ecdhkey),
                                              (const unsigned char *) eddsakey) != 0) {
-        zend_error(E_RECOVERABLE_ERROR, "crypto_sign_ed25519_pk_to_curve25519()");
+        zend_throw_exception(zend_ce_error, "crypto_sign_ed25519_pk_to_curve25519()", 0);
```
Comment (nikic) [inline libsodium.c]: Missing return

File: libsodium.c [hunk 9]
```diff
@@ -2343,18 +2557,18 @@ PHP_FUNCTION(sodium_compare)
...
-    if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "ss",
+    if (zend_parse_parameters(ZEND_NUM_ARGS(), "ss",
                               &buf1, &len1,
                               &buf2, &len2) == FAILURE) {
         return;
     }
     if (len1 != len2) {
-        zend_error(E_RECOVERABLE_ERROR, "compare(): arguments have different sizes");
+        zend_throw_exception(sodium_exception_ce, "compare(): arguments have different sizes", 0);
     } else if (len1 > SIZE_MAX) {
```
Comment (nikic) [inline libsodium.c]: Check no longer relevant

**Comment by nikic** [comment]
@paragonie-scott I noticed that some of the exceptions use `zend_ce_error`. Do we want to keep that or switch everything to `SodiumException`? If we want to keep it, I'll have to make further adjustments to make sure arguments are hidden for those two.

**Comment by nikic** [comment]
Another thing I noticed is that there are a lot of `>= SIZE_MAX` checks. If the `=` part of `>=` intentional?

**Comment by paragonie-scott** [comment]
> Do we want to keep that or switch everything to `SodiumException`?

Switch everything. I like `SodiumException` better.

> Another thing I noticed is that there are a lot of >= SIZE_MAX checks. Is the = part of >= intentional?

I think the goal there is to detect unsigned integer over/underflows? @jedisct1 would know more.

**Comment by jedisct1** [comment]
Awesome, thanks!

**Comment by jedisct1** [comment]
There are a lot of `>= STRSIZE_MAX` checks, not `>= SIZE_MAX`.

The goal is to avoid overflows, and still leave room for the final `\0`.

**Pull request merged**